clang/include/clang/CConv/ConstraintsGraph.h

```diff
@@ -12,125 +12,269 @@
 #ifndef _CONSTRAINTSGRAPH_H
 #define _CONSTRAINTSGRAPH_H
 
-#include <boost/graph/graph_traits.hpp>
-#include <boost/graph/adjacency_list.hpp>
 #include "clang/CConv/Constraints.h"
+#include "llvm/ADT/DirectedGraph.h"
+#include "llvm/Support/GraphWriter.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/BreadthFirstIterator.h"
 
-using namespace std;
+template<class DataType> struct DataEdge;
 
-template<class G>
-class BaseGraph {
+template<typename DataType, typename EdgeType = DataEdge<DataType>>
+class DataNode : public llvm::DGNode<DataNode<DataType, EdgeType>, EdgeType> {
 public:
-  typedef typename boost::graph_traits<G>::vertex_descriptor vertex_t;
-  typedef std::map<Atom*, vertex_t> VertexMapType;
-  G CG;
-protected:
-  VertexMapType AtomToVDMap;
+  typedef DataNode<DataType, EdgeType> NodeType;
+  typedef llvm::DGNode<NodeType, EdgeType> SuperType;
 
-  virtual vertex_t addVertex(Atom *A) {
-    if (AtomToVDMap.find(A) == AtomToVDMap.end()) {
-      auto Vidx = add_vertex(A, CG);
-      AtomToVDMap[A] = Vidx;
-    }
-    return AtomToVDMap[A];
+  DataNode() = delete;
+  DataNode(DataType D, EdgeType &E) : SuperType(E), Data(D) {}
+  explicit DataNode(DataType D) : SuperType(), Data(D) {}
+  DataNode(const DataNode &N) : SuperType(N), Data(N.Data) {}
+  DataNode(const DataNode &&N) : SuperType(std::move(N)), Data(N.Data) {}
+
+  ~DataNode() = default;
+
+  DataType getData() const { return Data; }
+
+  void addPred(EdgeType &E){ PredEdges.insert(&E); }
+  const llvm::SetVector<EdgeType *> &getPreds() { return PredEdges; }
+
+  bool isEqualTo(const DataNode &N) const { return this->Data == N.Data; }
+
+private:
+  DataType Data;
+
+  llvm::SetVector<EdgeType *> PredEdges;
+};
+
+template<typename Data, typename EdgeType>
+struct llvm::GraphTraits<DataNode<Data, EdgeType> *> {
```

Comment:
Don't understand this syntax. It looks as if you are defining a templated `struct`, but you are referring to the `llvm::` namespace, as if the `struct` is already defined. What's this doing?

Reply:
I think the term for this is partial template specialization. It lets you specialize a template given some constraints on the template arguments.   `GraphTraits` is a template struct defined by llvm that defines a sort of interface that you implement by specializing the struct for some particular class. In this case, I want to implement the interface for some a template class, so I need to do the partial specialization.